### ISSUE TITLE
Edit Locking Feature Missing

### ISSUE THREAD

**Issue by ThomasWaldmann** [moinwiki/moin#460]
**[Original report](https://bitbucket.org/thomaswaldmann/moin-2.0/issue/460) by RogerHaase (Bitbucket: [RogerHaase](https://bitbucket.org/RogerHaase), GitHub: [RogerHaase](https://github.com/RogerHaase)).**

----------------------------------------

In moin 1.9 there is an edit locking feature that warns users when someone else is editing the file. This feature is missing from moin 2.

**Comment by kwatsen** [comment]
I was about to open an issue for this...

Ideally the 2nd attempt to open the page for edit fails, but an alternate "solution" would be to fail the 2nd save attempt.  The different is in if editing is tracked versus if the `rev_number` is tracked (like ETags).  Of course, it means that the 2nd-user might lose their work, but they can save to an offline scratch pad and then recreate...

**Comment by ThomasWaldmann** [comment]
Failing late is imho even worse than blocking the edit. In 1.9 we have the option to choose the lock mode, but that doesn't really solve it either.

The minimum needed for fail-late (at save time) is to warn the user right from the beginning that someone else is also editing. If you would just let users edit in parallel without knowing and then fail the 2nd save, that would be frustrating to everybody doing bigger edits and having to time-consumingly merge them later manually in the other persons edit.

**Comment by ThomasWaldmann** [comment]
Note: and the only way to really solve this is to have a concurrent editor, we have another ticket about this, IIRC.

**Comment by kwatsen** [comment]
Yes, but if there is huge difference in effort, I'd take the "fail late" solution as a stop-gap.

One consideration for the "fail early" solution is the need to auto-release the lock after some timeout (an hour?).  But, if the user comes back later (perhaps seeing a "you're lock timed out" notice) and hit's "Ok", I believe that their attempt SHOULD be allowed to succeed, so long as no other change was made by another user in the meanwhile.  By extension, the "fail late" solution is seen to be part of the "fail early" solution.

Concurrent editing sounds like a distant nice-to-have...  :sigh:

**Comment by kwatsen** [comment]
Why is this issue labeled as a prio3/enhancement?  It seems more like a release-blocker to me...as wiki's are all about concurrent-editing...

**Comment by RogerHaase** [comment]
fixed by https://github.com/moinwiki/moin/pull/838
